### ISSUE TITLE
Custom view mode not used (D8 dev version)

### ISSUE THREAD

**Issue by jvandenrym** [drupal-fancy/view_mode_selector#17]
1. Create a custom view mode
2. Add view mode to content type
3. Edit a node of the content type and change the view mode from the view mode selector field
4. View mode value is saved but does not get applied on the node. Another view mode is used

**Comment by hawkeyetwolf** [comment]
Wherever that page is getting rendered, make sure it uses the "View Mode Selector" view mode, as that is how this module gets called in to render with the view mode you selected.

**Comment by preritmohan** [comment]
how?

**Comment by tbenice** [comment]
Right. E.G. nodes use the "full" view mode by default. To my knowledge there is no way to change this.

**Comment by RKopacz** [comment]
+1 on this behavior. I cannot apply a custom view mode to a node with this module. I have enabled the custom view mode for the content type. The View Mode Selector view mode has also been enabled for the content type in question. Not clear about @derekderaps instructions.

**Comment by preritmohan** [comment]
Hi All, I was able to get this to work with my custom view modes as well. I have used https://www.drupal.org/project/view_mode_selector module. After enabling the module... add a field of type "view mode selector" to the paragraph type.

Now the content type that has the "paragraphs" being embedded/referenced via entity referencing. In it's manage display select view mode as "View mode selector". That should do it. Attaching screenshot for reference. 

![image](https://user-images.githubusercontent.com/1901400/28641498-7e219df4-721d-11e7-9170-50f6cf31518a.png)

**Comment by preritmohan** [comment]
Also, custom view mode selectors won't work if Core module "Field Layout" is enabled.

Disable field layout module and then try view mode selector functionalities.

**Comment by jeromewiley** [comment]
Have the exact issue that jvandenrym is having. 

1. Create new View Mode in Drupal 8 site
2. Attach to my content type - news article
3. Install view_mode_selector module
4. Create new field of type view_mode_selector for news article
5. Configure default and special view modes (add a field, remove a field, etc.)
6. When I edit a node, and choose special, or choose default, only default view mode is used

**Comment by hawkeyetwolf** [comment]
@preritmohan, I suppose this approach doesn't work if you're trying to change the node's rendered output on page view. I think this module aims to allow the referenced entity of an ER field to control its output.

**Comment by tbenice** [comment]
Hey all, The issue seems to be with lines 14-16 of the .module file. It is not changing view modes if the selected view mode is not 'view mode selector'. Im not sure why that is there but removing it solves this issue, and referenced entities are still able to switch just fine.

It seems to me this check is not needed.

**Comment by yannickoo** [comment]
@tbenice the check makes sure we override the view mode only if the site builder wanted to have this rendered in the configured view mode. Otherwise it wouldn't be possible to force a specific view mode somewhere else.

**Comment by jrochate** [comment]
I'm controlling that with this simple function on a custom module:

function MYMODULE_entity_view_mode_alter(&$view_mode, Drupal\Core\Entity\EntityInterface $entity, $context) {
  // For nodes, change the view mode when it is page bundle and full view mode.
  if (($entity
    ->getEntityTypeId() == 'node') && ($entity->bundle() == 'page') && ($view_mode == 'full')) {
    $view_mode = 'view_mode_selector';
  }
}

**Comment by pjoulot** [comment]
I did kind of the same as jrochate but checking if the entity has a selector field and if it is the view mode I want.
It's kind of weird that you have to write code in order to make the module work.

Without any change, it is definitely keeping the default view mode instead of using the selected view mode at least on Drupal 9.

**Comment by makeaweli** [comment]
[Here's a module](https://gitlab.com/jason/noirlab-view-mode-selector-public) with similar functionality described by @jrochate and @pjoulot 

Definitely agree that this should be built-in to the module.

**Comment by yannickoo** [comment]
Hey there! I will check how we can fix this within the next days :)
